README.md
# kurly

[![Build Status](https://travis-ci.org/davidjpeacock/kurly.svg?branch=master)](https://travis-ci.org/davidjpeacock/kurly) [![Snap Status](https://build.snapcraft.io/badge/letozaf/kurly.svg)](https://build.snapcraft.io/user/letozaf/kurly)

kurly is an alternative to the widely popular curl program.

kurly is designed to operate in a similar manner to curl, with select features.
Notably, kurly is not aiming for feature parity, but common flags and mechanisms
particularly within the HTTP(S) realm are to be expected.

The current authors are not security experts, but want to contribute to the fledging
movement of replacing key tools and services with equivalents based on modern
and safe languages.  We recognize that people are fallible (including
ourselves), and for this reason believe we need all the help we can get.

Several languages exist which could be used to fulfill our goal, but in this case
we picked Golang.

## Installation

**Pre-requisite: Golang version 1.7.4 or higher.**

From source you can simply:

`go get github.com/davidjpeacock/kurly`

## OS Package

`kurly` can be installed through package management systems on the following platforms:

* Arch Linux via Arch User Repos
  + For stable versions : `pacaur -S kurly` or `yaourt -S kurly`
  + For tip/development versions : `pacaur -S kurly-git` or `yaourt -S kurly-git` 
* Linux, using the [snap package](https://snapcraft.io/docs/core/install) - `snap install kurly`
* Linux, installing the snap package from within the desktop app store: go to [https://snapcraft.io/kurly](https://snapcraft.io/kurly?mkt_tok=eyJpIjoiTmpBd056UmtZV1U1TVRrMSIsInQiOiJJN0U4RWNUSFN2NjZNV3hjOTFBaGpoWnJnRkdJWnFZVnUxeFE0SzJMYnU3Sit5cnh1anFLNkpMVUhOSjhBaENIN0d1T1FiUFdSMmVWR28zM3VqUHZLNHdsN0daVHhpdjFXNVRtMEJweDdYajVxT1FTSjEwdTZJekxpRjBTR1wvbGMifQ%3D%3D)
  clicking the install button on this landing page will install Kurly.
* Linux x86 64 via [RPM](https://github.com/davidjpeacock/kurly/releases/download/v1.2.1/kurly-1.2.1-0.x86_64.rpm) - `yum install kurly-1.2.1-0.x86_64.rpm`

*If you're a package maintainer and you have prepared kurly for your OS of choice, please
PR this section.*

## Binary download

Binaries are provided for the following platforms:

* [Linux amd64](https://github.com/davidjpeacock/kurly/releases/download/v1.2.1/kurly-linux-amd64-v1.2.1.tar.gz)
* [Linux arm](https://github.com/davidjpeacock/kurly/releases/download/v1.2.1/kurly-linux-arm-v1.2.1.tar.gz)
* [Mac OS X amd64](https://github.com/davidjpeacock/kurly/releases/download/v1.2.1/kurly-osx-amd64-v1.2.1.tar.gz)
* [Windows amd64](https://github.com/davidjpeacock/kurly/releases/download/v1.2.1/kurly-windows-amd64-v1.2.1.zip)

## Usage

See `kurly --help` for usage information.

## Examples

Verbose output, showing headers
```
$ kurly -v https://httpbin.org/ip
*   Trying 23.23.171.5...
* TCP_NODELAY set
* Connected to httpbin.org (23.23.171.5) port 443 (#0)
* APLN, server accepted to use http/1.1
* TLSv1.2, TLS Handshake finished
* SSL connection using TLSv1.2 / ECDHE-RSA-AES-128-GCM-SHA256
* Server certificate:
*  subject: CN=httpbin.org
*  start date: Thu, 11 Jan 2018 23:37:29 UTC
*  expire date: Wed, 11 Apr 2018 23:37:29 UTC
*  issuer: C=US; O=Let's Encrypt; CN=Let's Encrypt Authority X3
*  SSL certificate verify ok.
> GET /ip HTTP/1.1
> User-Agent [Kurly/1.2.1]
> Accept [*/*]
> Host [httpbin.org]
< HTTP/1.1 200 OK
< Server [meinheld/0.6.1]
< Content-Type [application/json]
< Access-Control-Allow-Credentials [true]
< Content-Length [33]
< Via [1.1 vegur]
< Connection [keep-alive]
< Date [Mon, 12 Mar 2018 19:18:11 GMT]
< Access-Control-Allow-Origin [*]
< X-Powered-By [Flask]
< X-Processed-Time [0]
[<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<] 33 B/33 B
{
  "origin": "43.122.23.223"
}
```

Download file, preserving remote filename, timestamp, and following redirects
```
$ kurly -R -O -L http://cdimage.debian.org/debian-cd/current/amd64/iso-cd/debian-8.7.1-amd64-netinst.iso
[<<<<<<                                ] 41.2 MB/260 MB
```

Upload file
```
$ kurly -T ~/Downloads/image.jpeg https://httpbin.org/put
```

Posting elements with -d
```
$ kurly -d bingo=bongo https://httpbin.org/post
```

## Roadmap

Succinctly, we're planning to cover all curl-like features relevant to HTTP(S), and would
love you to help.

## Contributing

Bug reports, feature requests, and pull requests are all welcome.  Thank you!

Please see [CONTRIBUTING.md](https://github.com/davidjpeacock/kurly/blob/master/CONTRIBUTING.md) for details of how to work with us.

## Maintainers

kurly is brought to you and maintained by:

* [Akilan Elango](https://github.com/aki237)
* [Al S-M](https://github.com/alsm)
* [David J Peacock](https://github.com/davidjpeacock)

## License

kurly is Copyright (c) 2017-2018 David J Peacock and Al S-M, and is published under the Apache 2.0 license.


CHANGELOG.md
# Changelog
This file contains a record of all non-trivial changes to kurly

## [Unreleased]

## [1.2.1] 20180312

### Added
* Improved verbosity
* TLS Verbosity
* Support for insecure HTTPS
* Added man page
* Behind-the-scenes refactor for future maintenance
* Handle multiple URLs
* Snap installation via desktop UI

## [1.1.0] 20171229

### Added
* Resume transfer from offset
* Cookie and cookie jar support
* Send HTTP multipart post data
* Installation for Linux via snap package
* Installation for Arch Linux via Arch User Repos

## [1.0.0] 20170501

Initial public release


main_test.go
package main

import (
	"net/http"
	"strings"
	"testing"
)

func TestSetHeaders(t *testing.T) {
	t.Log("Testing setHeaders()... (expecting Kurly/1.0)")

	header := []string{"User-Agent: Kurly/1.0"}
	req, err := http.NewRequest("GET", "http://url.com/", nil)
	if err != nil {
		panic(err)
	}

	setHeaders(req, header)

	if len(req.Header) > 0 {
		for _, v := range req.Header {
			userAgentValue := strings.Join(v, "")
			if userAgentValue != "Kurly/1.0" {
				t.Errorf("Expected Kurly/1.0, but got %s", userAgentValue)
			}
		}
	} else {
		t.Error("setHeaders() set no header")
	}
}

/*** Need to investigate `go test -race` which causes this to fail and break CI
func TestMaxTime(t *testing.T) {
	t.Log("Testing maxTime()... (expecting no timeout)")

	maxTime(1)
	time.Sleep(500 * time.Millisecond)
}
*/


trace.go
package main

import (
	"crypto/tls"
	"fmt"
	"net"
	"net/http"
	"net/http/httptrace"
	"strings"
)

type tracerStruct struct {
	req         *http.Request
	redirects   int
	currentHost string
}

func (ts *tracerStruct) DNSStart(dnsinfo httptrace.DNSStartInfo) {
	ts.currentHost = dnsinfo.Host
}

func (ts *tracerStruct) ConnectStart(network, addr string) {
	hs, _, _ := net.SplitHostPort(addr)
	Status.Printf("   Trying %s...\n", hs)
}

func (ts *tracerStruct) WroteHeaders() {
	fmt.Fprintln(Outgoing, ts.req.Method, ts.req.URL.Path, ts.req.Proto)
	for k, v := range ts.req.Header {
		fmt.Fprintln(Outgoing, k, v)
	}
}

func (ts *tracerStruct) ConnectDone(network, addr string, err error) {
	Status.Println(" TCP_NODELAY set")
	hs, port, _ := net.SplitHostPort(addr)
	Status.Printf(" Connected to %s (%s) port %s (#%d)%s\n", ts.currentHost, hs, port, ts.redirects, ts.req.RequestURI)
	ts.redirects += 1
}

func (ts *tracerStruct) GotConn(cinfo httptrace.GotConnInfo) {
	if cinfo.Reused {
		hs, port, _ := net.SplitHostPort(cinfo.Conn.RemoteAddr().String())
		Status.Printf(" Re-using existing connection! (#%d) with host %s\n", ts.redirects-1, ts.currentHost)
		Status.Printf(" Connected to %s (%s) port %s (#%d)%s\n", ts.currentHost, hs, port, ts.redirects, ts.req.RequestURI)
	}
}

func (ts *tracerStruct) TLSHandshakeDone(cstate tls.ConnectionState, err error) {
	if !cstate.HandshakeComplete {
		Status.Printf(" TLS Handshake not completed")
		return
	}
	var tlsversion, ciphersuite string
	switch cstate.Version {
	case tls.VersionSSL30:
		tlsversion = "SSLv3"
	case tls.VersionTLS10:
		tlsversion = "TLSv1.0"
	case tls.VersionTLS11:
		tlsversion = "TLSv1.1"
	case tls.VersionTLS12:
		tlsversion = "TLSv1.2"
	}

	ciphersuite = getCipherSuiteString(cstate.CipherSuite)
	Status.Printf(" APLN, server accepted to use %s", cstate.NegotiatedProtocol)
	Status.Printf(" %s, TLS Handshake finished", tlsversion)
	Status.Printf(" SSL connection using %s / %s", tlsversion, ciphersuite)
	if len(cstate.PeerCertificates) > 0 {
		cert := cstate.PeerCertificates[0]
		Status.Println(" Server certificate:")
		Status.Printf("  subject: CN=%s\n", cert.Subject.CommonName)
		Status.Printf("  start date: %s\n", cert.NotBefore.Format("Mon, 02 Jan 2006 15:04:05 MST"))
		Status.Printf("  expire date: %s\n", cert.NotAfter.Format("Mon, 02 Jan 2006 15:04:05 MST"))
		Status.Printf("  issuer: C=%s; O=%s; CN=%s\n",
			strings.Join(cert.Issuer.Country, " "),
			strings.Join(cert.Issuer.Organization, " "),
			cert.Issuer.CommonName)
		Status.Printf("  SSL certificate verify ok.\n")
	}
}

func NewClientTraceForRequest(req *http.Request) *httptrace.ClientTrace {
	if req == nil {
		Status.Fatal("cannot do a verbose trace for a empty request")
	}

	ts := &tracerStruct{req: req}

	return &httptrace.ClientTrace{
		ConnectStart:     ts.ConnectStart,
		ConnectDone:      ts.ConnectDone,
		WroteHeaders:     ts.WroteHeaders,
		GotConn:          ts.GotConn,
		DNSStart:         ts.DNSStart,
		TLSHandshakeDone: ts.TLSHandshakeDone,
	}
}

func getCipherSuiteString(num uint16) string {
	switch num {
	case tls.TLS_RSA_WITH_RC4_128_SHA:
		return "RSA-RC4-128-SHA"
	case tls.TLS_RSA_WITH_3DES_EDE_CBC_SHA:
		return "RSA-3DES-EDE-CBC-SHA"
	case tls.TLS_RSA_WITH_AES_128_CBC_SHA:
		return "RSA-AES-128-CBC-SHA"
	case tls.TLS_RSA_WITH_AES_256_CBC_SHA:
		return "RSA-AES-256-CBC-SHA"
	case tls.TLS_RSA_WITH_AES_128_CBC_SHA256:
		return "RSA-AES-128-CBC-SHA256"
	case tls.TLS_RSA_WITH_AES_128_GCM_SHA256:
		return "RSA-AES-128-GCM-SHA256"
	case tls.TLS_RSA_WITH_AES_256_GCM_SHA384:
		return "RSA-AES-256-GCM-SHA384"
	case tls.TLS_ECDHE_ECDSA_WITH_RC4_128_SHA:
		return "ECDHE-ECDSA-RC4-128-SHA"
	case tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA:
		return "ECDHE-ECDSA-AES-128-CBC-SHA"
	case tls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA:
		return "ECDHE-ECDSA-AES-256-CBC-SHA"
	case tls.TLS_ECDHE_RSA_WITH_RC4_128_SHA:
		return "ECDHE-RSA-RC4-128-SHA"
	case tls.TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA:
		return "ECDHE-RSA-3DES-EDE-CBC-SHA"
	case tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA:
		return "ECDHE-RSA-AES-128-CBC-SHA"
	case tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA:
		return "ECDHE-RSA-AES-256-CBC-SHA"
	case tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256:
		return "ECDHE-ECDSA-AES-128-CBCSHA256"
	case tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256:
		return "ECDHE-RSA-AES-128-CBC-SHA256"
	case tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256:
		return "ECDHE-RSA-AES-128-GCM-SHA256"
	case tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256:
		return "ECDHE-ECDSA-AES-128-GCM-SHA256"
	case tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384:
		return "ECDHE-RSA-AES-256-GCM-SHA384"
	case tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384:
		return "ECDHE-ECDSA-AES-256-GCM-SHA384"
	case tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305:
		return "ECDHE-RSA-CHACHA20-POLY1305"
	case tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305:
		return "ECDHE-ECDSA-CHACHA20-POLY1305"
	}
	return "FALLBACK-SCSV"
}


main.go
package main

import (
	"bytes"
	"crypto/tls"
	"encoding/base64"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"mime/multipart"
	"net/http"
	"net/http/httptrace"
	"net/textproto"
	"net/url"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/aki237/nscjar"
	"github.com/alsm/ioprogress"
	"github.com/davidjpeacock/cli"
)

var (
	client   = http.Client{}
	Status   = log.New(os.Stderr, "*", 0)
	Incoming io.Writer
	Outgoing io.Writer
)

const (
	version string = "1.2.1"
)

type LogWriter struct {
	*log.Logger
}

func (l *LogWriter) Write(p []byte) (n int, err error) {
	l.Print(string(p))
	return len(p), nil
}

func init() {
	cli.VersionFlag = cli.BoolFlag{
		Name:  "version, V",
		Usage: "print the version",
	}
	Incoming = &LogWriter{Logger: log.New(ioutil.Discard, "< ", 0)}
	Outgoing = &LogWriter{Logger: log.New(ioutil.Discard, "> ", 0)}
}

func main() {
	var opts Options

	app := cli.NewApp()
	app.Name = "kurly"
	app.Usage = "[options] URL"
	app.Version = version

	opts.getOptions(app)

	app.Action = func(c *cli.Context) error {
		if c.NArg() == 0 {
			cli.ShowAppHelp(c)
			os.Exit(0)
		}

		err := opts.BuildCommonOptions(c)
		if err != nil {
			return err
		}

		for _, uri := range c.Args() {
			err := fetchUrl(uri, opts, c)
			if err != nil {
				fmt.Fprintf(os.Stderr, "kurly : %s\n", err)
			}
		}
		return nil
	}

	app.Run(os.Args)
}

func fetchUrl(target string, opts Options, c *cli.Context) error {
	var remote *url.URL
	var err error
	var body io.Reader

	err = opts.BuildTargetSpecificOptions(target, body)
	if err != nil {
		return err
	}

	client.CheckRedirect = opts.checkRedirect
	if opts.insecure {
		client.Transport = &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
		}
	}

	if remote, err = url.Parse(target); err != nil {
		return fmt.Errorf("Error: %s does not parse correctly as a URL", target)
	}

	if remote.Scheme == "" {
		remote.Scheme = "http"
		remote, _ = url.Parse(remote.String())
	}

	outputFile := opts.openOutputFile()

	continueAtInt := uint64(0)
	if opts.continueAt != "" {
		if opts.continueAt == "-" {
			fileInfo, err := outputFile.Stat()
			if err != nil {
				return fmt.Errorf("unable to set content range automatically from file; %s", err)
			}
			continueAtInt = uint64(fileInfo.Size())
		} else {
			continueAtInt, err = strconv.ParseUint(opts.continueAt, 10, 64)
			if err != nil {
				return fmt.Errorf("unable to create http request; expected a valid positive number for continue-at option")
			}
		}
	}

	req, err := http.NewRequest(opts.method, target, body)
	if err != nil {
		Status.Fatalf("Error: unable to create http %s request; %s\n", opts.method, err)
	}

	if opts.verbose {
		req = req.WithContext(httptrace.WithClientTrace(req.Context(), NewClientTraceForRequest(req)))
	}

	// Seek to given offset of the file and set the "Range" header
	if continueAtInt > 0 {
		if opts.outputFilename != "" {
			_, err = outputFile.Seek(int64(continueAtInt), 0)
			if err != nil {
				Status.Fatalf("Error: seek in the given output file; %s\n", err)
			}
		}
		req.Header.Set("Range", fmt.Sprintf("bytes=%d-", continueAtInt))
	}
	req.Header.Set("User-Agent", opts.agent)
	if opts.user != "" {
		req.Header.Set("Authorization", "Basic "+encodeToBase64(opts.user))
	}
	req.Header.Set("Accept", "*/*")
	req.Header.Set("Host", remote.Host)
	if body != nil {
		switch b := body.(type) {
		case *os.File:
			fi, err := b.Stat()
			if err != nil {
				Status.Fatalf("Unable to get file stats for %v\n", opts.fileUpload)
			}
			req.ContentLength = fi.Size()
			req.Header.Set("Content-Length", strconv.FormatInt(fi.Size(), 10))
		case *ioprogress.Reader:
			req.ContentLength = b.Size
			req.Header.Set("Content-Length", strconv.FormatInt(b.Size, 10))
		case *bytes.Buffer:
			req.Header.Set("Content-Length", strconv.FormatInt(int64(b.Len()), 10))
		}
	}
	setHeaders(req, opts.headers)
	setCookieHeader(req, opts.cookie)

	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if continueAtInt > 0 && resp.StatusCode == 416 {
		return fmt.Errorf("unable to get URL; %s\n", "Either the server doesn't support ranges or an invalid range is passed")
	}

	fmt.Fprintf(Incoming, "%s %s\n", resp.Proto, resp.Status)

	for k, v := range resp.Header {
		fmt.Fprintln(Incoming, k, v)
	}

	if !opts.head {
		if !opts.silent {
			progressR := &ioprogress.Reader{
				Reader: resp.Body,
				Size:   resp.ContentLength,
				DrawFunc: ioprogress.DrawTerminalf(os.Stderr, func(progress, total int64) string {
					return fmt.Sprintf(
						"%s %s",
						(ioprogress.DrawTextFormatBarWithIndicator(40, '<'))(progress, total),
						ioprogress.DrawTextFormatBytes(progress, total))
				}),
			}
			if _, err = io.Copy(outputFile, progressR); err != nil {
				return fmt.Errorf("failed to copy URL content; %s", err)
			}
		}
		if opts.silent {
			if _, err = io.Copy(outputFile, resp.Body); err != nil {
				return fmt.Errorf("failed to copy URL content; %s", err)
			}
		}
	}

	if opts.outputFilename != "" {
		outputFile.Close()
	}

	if rTime := resp.Header.Get("Last-Modified"); opts.remoteTime && rTime != "" {
		if t, err := time.Parse("Mon, 02 Jan 2006 15:04:05 MST", rTime); err == nil {
			os.Chtimes(opts.outputFilename, t, t)
		}
	}

	if opts.cookieJar != "" && len(resp.Cookies()) > 0 {
		cookies := resp.Cookies()
		for _, val := range cookies {
			if val.Domain == "" {
				u, err := resp.Location()
				if err != nil {
					u = req.URL
				}

				val.Domain = u.Hostname()
			}
		}
		saveCookies(cookies, opts.cookieJar)
	}
	return nil
}

func setHeaders(r *http.Request, h []string) {
	for _, header := range h {
		hParts := strings.Split(header, ": ")
		switch len(hParts) {
		case 0:
			//surely not
		case 1:
			//must be an empty Header or a delete
			switch {
			case strings.HasSuffix(header, ";"):
				r.Header.Set(strings.TrimSuffix(header, ";"), "")
			case strings.HasSuffix(header, ":"):
				r.Header.Del(strings.TrimSuffix(header, ":"))
			default:
			}
		case 2:
			//standard expected
			r.Header.Set(hParts[0], hParts[1])
		default:
			//more than expected, use first element as Header name
			//and rejoin the rest as header content
			r.Header.Set(hParts[0], strings.Join(hParts[1:], ": "))
		}
	}
}

func writeToMultipart(w *multipart.Writer, key string, field Field) error {
	if field.IsFile {
		file, err := os.Open(field.Value)
		if err != nil {
			return err
		}
		defer file.Close()

		h := make(textproto.MIMEHeader)
		h.Set("Content-Disposition",
			fmt.Sprintf(`form-data; name="%s"; filename="%s"`,
				key, field.Filealias))
		h.Set("Content-Type", "application/octet-stream")
		if field.Type != "" {
			h.Set("Content-Type", field.Type)
		}

		fw, err := w.CreatePart(h)
		if err != nil {
			return err
		}
		if _, err = io.Copy(fw, file); err != nil {
			return err
		}
	} else {
		h := make(textproto.MIMEHeader)
		h.Set("Content-Disposition",
			fmt.Sprintf(`form-data; name="%s"`, key))
		if field.Type != "" {
			h.Set("Content-Type", field.Type)
		}
		wr, err := w.CreatePart(h)
		if err != nil {
			return err
		}
		_, err = wr.Write([]byte(field.Value))
		if err != nil {
			return err
		}

	}
	return nil
}

func setCookieHeader(r *http.Request, arg string) {
	// according to cURL man pages and operations, if the cookie string passed has
	// a "=" in it, it means that is a valid cookie. Else it will search for the
	// cookie jar file of the name and tries to read cookies from it
	if strings.Contains(arg, "=") {
		r.Header.Set("Cookie", arg)
		return
	}

	// If the specified cookie string is invalid and the the file of the same name
	// doesn't exist, cURL doesn't throw any error, rather it just sets the cookie
	// to null
	cookieFile, err := os.Open(arg)
	if err != nil {
		return
	}

	p := nscjar.Parser{}

	cookies, err := p.Unmarshal(cookieFile)
	if err != nil {
		return
	}

	// reference time for checking the expiry of a cookie from the file
	t := time.Now()

	for _, val := range cookies {
		// Skip if cookie is secure and request isn't
		if val.Secure && r.URL.Scheme != "https" {
			continue
		}

		// Skip if the cookie has expired
		if val.Expires.Unix() <= t.Unix() {
			continue
		}

		host := r.URL.Hostname()
		// Domain matching (according to rfc6265::Section-5.1.3)
		if strings.ToLower(val.Domain) != strings.ToLower(host) {
			suffix := val.Domain
			if !strings.HasPrefix(val.Domain, ".") {
				suffix = "." + suffix
			}
			if !strings.HasSuffix(host, suffix) {
				continue
			}
		}

		// Path matching (according to rfc6265::Section-5.1.4)
		uri, _ := url.Parse("http://" + val.Domain + val.Path)
		if strings.HasPrefix(r.URL.Path, uri.Path) && r.URL.Path != uri.Path {
			continue
		}

		r.AddCookie(val)
	}
}

func saveCookies(cs []*http.Cookie, filename string) {
	f, err := os.OpenFile(filename, os.O_CREATE|os.O_RDWR, 0666)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Warning : unable to save the cookie to the file : %s\n", err)
		return
	}

	jar := nscjar.NewCookieJar()

	// get the already stored cookies if exists in file.
	p := nscjar.Parser{}
	if cookies, err := p.Unmarshal(f); err == nil {
		jar.AddCookies(cookies...)
	}

	jar.AddCookies(cs...)

	_, err = f.Seek(0, 0)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Warning : unable to save the cookie to the file : %s\n", err)
		return
	}

	err = jar.Marshal(f)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Warning : unable to save the cookie to the file : %s\n", err)
	}
	err = f.Close()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Warning : unable to save the cookie to the file : %s\n", err)
	}
}

func maxTime(maxTime uint) {
	go func() {
		<-time.After(time.Duration(maxTime) * time.Second)
		Status.Fatalf("Error: Maximum operation time of %d seconds expired, aborting\n", maxTime)
	}()
}

func encodeToBase64(a string) string {
	return base64.StdEncoding.EncodeToString([]byte(a))
}
